src/library_pthread.js

```diff
@@ -1172,7 +1137,56 @@ var LibraryPThread = {
 #if PTHREADS_PROFILING
     _emscripten_set_thread_name_js(threadId|0, name|0);
 #endif
-  }
+  },
+
+  emscripten_proxy_to_main_thread_js: function() {
+    // Arguments are:
+    //   * the index of the function
+    //   * the call args
+    //   * whether it is sync
+    var index = arguments[0];
+    var sync = arguments[arguments.length - 1];
+    // The serialization buffer contains the number of call params, and then
+    // all the args here.
+    // We also pass 'sync' to C separately, since C needs to look at it.
+    // The buffer remains alive until receiveOnMainThread frees it.
+    var numCallArgs = arguments.length - 2;
+    // Allocate a buffer, which will be copied by the C code.
+    var stack = stackSave();
+    var buffer = stackAlloc(numCallArgs * 8);
+    for (var i = 0; i < numCallArgs; i++) {
+      HEAPF64[(buffer >> 3) + i] = arguments[1 + i];
+    }
+    var ret = _emscripten_run_in_main_runtime_thread_js(index, numCallArgs, buffer, sync);
+    stackRestore(stack);
+    return ret;
+  },
+
+  emscripten_receive_on_main_thread_js__deps: ['emscripten_proxy_to_main_thread_js'],
+  emscripten_receive_on_main_thread_js: function(index, numCallArgs, buffer) {
+    // Avoid garbage by reusing a single JS array for call arguments.
+    if (!_emscripten_receive_on_main_thread_js.callArgs) {
+      _emscripten_receive_on_main_thread_js.callArgs = [];
+    }
+    var callArgs = _emscripten_receive_on_main_thread_js.callArgs;
+    callArgs.length = numCallArgs;
+    for (var i = 0; i < numCallArgs; i++) {
+      callArgs[i] = HEAPF64[(buffer >> 3) + i];
+    }
+    var sync = HEAPF64[(buffer >> 3) + 2 + numCallArgs];
+    // Proxied JS library funcs are encoded as positive values, and
+    // EM_ASMs as negative values (see include_asm_consts)
+    var func;
+    if (index > 0) {
+      func = proxiedFunctionTable[index];
+    } else {
+      func = ASM_CONSTS[-index - 1];
+    }
+#if ASSERTIONS
+    assert(func.length == numCallArgs);
+#endif
+    return func.apply(null, callArgs);
```

Comment:
Still really skeptical about `func.apply()` and the dynamic `array.length = x;` patterns, I have not seen good JS engine behavior from them - but looks great for now, I can come back to micro-optimize if profiling data shows something bad here.

Reply:
Yeah, worth benchmarking to make sure. But I believe JS engines have gotten a lot better at this kind of thing in my experience, and it is shorter code.